components/p3a/star_randomness_meta.cc

```diff
@@ -0,0 +1,270 @@
+                  "attestation";
+    AttestServer(false);
+    return false;
+  }
+  return true;
+}
+
+void StarRandomnessMeta::RequestServerInfo() {
+  rnd_server_info_ = nullptr;
+
+  if (!config_->disable_star_attestation && approved_cert_ == nullptr) {
+    AttestServer(true);
+    return;
+  }
+
+  if (!has_used_cached_info) {
+    // Using cached server info, if available.
+    base::Time saved_next_epoch_time =
+        local_state_->GetTime(kNextEpochTimePrefName);
+    if (saved_next_epoch_time > base::Time::Now()) {
+      int saved_epoch = local_state_->GetInteger(kCurrentEpochPrefName);
+      std::string saved_pk = local_state_->GetString(kCurrentPKPrefName);
+
+      rnd_server_info_ = std::make_unique<RandomnessServerInfo>(
+          static_cast<uint8_t>(saved_epoch), saved_next_epoch_time,
+          DecodeServerPublicKey(&saved_pk));
+      VLOG(2) << "StarRandomnessMeta: using cached server info";
+      info_callback_.Run(rnd_server_info_.get());
+      has_used_cached_info = true;
+      return;
+    }
+  }
+
+  rnd_server_info_ = nullptr;
+  auto resource_request = std::make_unique<network::ResourceRequest>();
+  resource_request->url = GURL(config_->star_randomness_host + "/info");
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(resource_request), GetRandomnessRequestAnnotation());
+  url_loader_->SetURLLoaderFactoryOptions(
+      network::mojom::kURLLoadOptionSendSSLInfoWithResponse);
+  url_loader_->DownloadToString(
+      url_loader_factory_.get(),
+      base::BindOnce(&StarRandomnessMeta::HandleServerInfoResponse,
+                     base::Unretained(this)),
+      kMaxInfoResponseSize);
+}
+
+RandomnessServerInfo* StarRandomnessMeta::GetCachedRandomnessServerInfo() {
+  return rnd_server_info_.get();
+}
+
+void StarRandomnessMeta::AttestServer(bool make_info_request_after) {
+  if (attestation_pending_) {
+    return;
+  }
+  attestation_pending_ = true;
+  approved_cert_ = nullptr;
+
+  VLOG(2) << "StarRandomnessMeta: starting attestation";
+  GURL attestation_url = GURL(config_->star_randomness_host + "/attestation");
+  nitro_utils::RequestAndVerifyAttestationDocument(
+      attestation_url, url_loader_factory_.get(),
+      base::BindOnce(&StarRandomnessMeta::HandleAttestationResult,
+                     base::Unretained(this), make_info_request_after));
+}
+
+void StarRandomnessMeta::HandleAttestationResult(
+    bool make_info_request_after,
+    scoped_refptr<net::X509Certificate> approved_cert) {
+  if (approved_cert == nullptr) {
+    LOG(ERROR) << "StarRandomnessMeta: attestation failed";
+    attestation_pending_ = false;
+    if (make_info_request_after) {
+      ScheduleServerInfoRetry();
+    }
+    return;
+  }
+  approved_cert_ = approved_cert;
+  attestation_pending_ = false;
+  VLOG(2) << "StarRandomnessMeta: attestation succeeded";
+  if (make_info_request_after) {
+    RequestServerInfo();
+  }
+}
+
+void StarRandomnessMeta::ScheduleServerInfoRetry() {
```

Comment:
would be great to switch our backoffs to `net::BackoffEntry`, probably in a follow-up

Reply:
agreed, looks promising